sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,244 @@
+
+/**
+ * Generic padctrl DIF return codes
+ */
+typedef enum dif_padctrl_result {
+  /**
+   * Operation succeeded.
+   */
+  kDifPadctrlOk = 0,
+  /**
+   * Operation failed due to an unspecified error. It cannot be recovered or
+   * retried. */
+  kDifPadctrlError = 1,
+  /**
+   * Operation failed due to a bad argument and can be retried. No register
+   * reads or writes have been performed.
+   */
+  kDifPadctrlBadArg = 2,
+} dif_padctrl_result_t;
+
+/**
+ * Pad kinds.
+ *
+ * A pad can be an MIO or DIO pad
+ */
+typedef enum dif_padctrl_pad_kind {
+  /**
+   * MIO (Multiplexed Input/Output) pad. Connected via pinmux.
+   */
+  kDifPadctrlPadMio = 0,
+  /**
+   * DIO (Dedicated Input/Output) pad. Connected directly to peripheral inputs
+   * and outputs, bypassing pinmux.
+   */
+  kDifPadctrlPadDio,
+} dif_padctrl_pad_kind_t;
+
+/**
+ * Pad index.
+ *
+ * Combined with `dif_padctrl_pad_kind_t` to refer to a specific pad.
+ */
+typedef uint32_t dif_padctrl_pad_index_t;
+
+/**
+ * Padctrl instance state.
+ *
+ * Padctrl persistent data that is required by the Padctrl API.
+ */
+typedef struct dif_padctrl {
+  mmio_region_t base_addr; /**< Padctrl base address. */
+} dif_padctrl_t;
+
+/**
+ * Initialises an instance of Padctrl.
+ *
+ * Information that must be retained, and is required to program Padctrl, shall
+ * be stored in @p padctrl.
+ *
+ * @param base_addr Base address of an isntance of the Padctrl IP block
+ * @param padctrl Padctrl state data
+ * @return dif_padctrl_init_result_t.
+ */
+dif_padctrl_result_t dif_padctrl_init(mmio_region_t base_addr,
+                                      dif_padctrl_t *padctrl);
+
+/**
+ * Padctrl reset routine error codes.
+ */
+typedef enum dif_padctrl_reset_result {
+  kDifPadctrlResetOk = kDifPadctrlOk,
+  kDifPadctrlResetError = kDifPadctrlError,
+  kDifPadctrlResetBadArg = kDifPadctrlBadArg,
+  /**
+   * Peripheral is locked and cannot be reset.This is not recoverable.
+   */
+  kDifPadctrlResetLocked,
+} dif_padctrl_reset_result_t;
+
+/**
+ * Resets padctrl by writing a given set of attributes to every pad
+ *
+ * @param padctrl Padctrl state data
+ * @param reset_attrs Attributes to write to every pad
+ * @return dif_padctrl_result_t
+ */
+dif_padctrl_reset_result_t dif_padctrl_reset(dif_padctrl_t *padctrl,
+                                             dif_padctrl_attr_t reset_attrs);
+
+/**
+ * Locks the padctrl registers.
+ *
+ * After the registers have been locked, they can only be unlocked by the
+ * system reset.
+ *
+ * @param padctrl Padctrl state data.
+ * @return dif_padctrl_result_t.
+ */
+dif_padctrl_result_t dif_padctrl_lock(dif_padctrl_t *padctrl);
+
+/**
+ * Padctrl attribute change routines error codes.
+ */
+typedef enum dif_padctrl_change_attr_result {
+  kDifPadctrlChangeOk = kDifPadctrlOk,
+  kDifPadctrlChangeError = kDifPadctrlError,
+  kDifPadctrlChangeBadArg = kDifPadctrlBadArg,
+  /**
+   * Cannot perform attribute change as the padctrl registers are locked
+   */
```

Comment:
Maybe make this comment identical to the one for `reset_result`?

Also, I think these variants need to technically be named `ChangeAttr`.